### PR TITLE
Add more search options to Beatmap Listing

### PR DESCRIPTION
**Pull request by EVAST9919**
Part of #9030.

Screenshot_2

### CODE REVIEW

File: osu.Game/Online/API/Requests/SearchBeatmapSetsRequest.cs [hunk 1]
```diff
@@ -66,6 +80,15 @@ protected override WebRequest CreateWebRequest()
 
             req.AddParameter("sort", $"{SortCriteria.ToString().ToLowerInvariant()}_{directionString}");
 
+            if (Extra != null && Extra.Any())
+                req.AddParameter("e", string.Join(".", Extra.Select(e => e.ToString().ToLowerInvariant())));
```
Comment (turbedi) [inline osu.Game/Online/API/Requests/SearchBeatmapSetsRequest.cs:84]:
```suggestion
                req.AddParameter("e", string.Join('.', Extra.Select(e => e.ToString().ToLowerInvariant())));
```

File: osu.Game/Online/API/Requests/SearchBeatmapSetsRequest.cs [hunk 2]
```diff
@@ -66,6 +80,15 @@ protected override WebRequest CreateWebRequest()
...
+
+            if (Ranks != null && Ranks.Any())
+                req.AddParameter("r", string.Join(".", Ranks.Select(r => r.ToString())));
```
Comment (turbedi) [inline osu.Game/Online/API/Requests/SearchBeatmapSetsRequest.cs:87]:
```suggestion
                req.AddParameter("r", string.Join('.', Ranks.Select(r => r.ToString())));
```

File: osu.Game/Overlays/BeatmapListing/SearchRank.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.ComponentModel;
+
+namespace osu.Game.Overlays.BeatmapListing
+{
+    public enum SearchRank
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/SearchRank.cs:8]: Skeptical this needs to exist alongside `ScoreRank`. Would just use that instead and provide a way to override `ScoreRank`'s `Description`s.

File: osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        protected abstract class MultipleSelectionFilter : FillFlowContainer<MultipleSelectionFilterTabItem>, IHasCurrentValue<List<T>>
+        {
+            private readonly BindableWithCurrent<List<T>> current = new BindableWithCurrent<List<T>>();
+
+            public Bindable<List<T>> Current
+            {
+                get => current.Current;
+                set => current.Current = value;
+            }
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs:34]: `Bindable<List<T>>` is not done anywhere. `BindableList<T>` exists in framework and has proper handling of callbacks on add/remove/replace/move/reset operations. You wouldn't have to do this whole rigamarole of reconstructing the entire list to trigger a value change if you used it.

File: osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                        selectedValues.Add(item.Value);
+                }
+
+                Current.Value = selectedValues;
+            }
+        }
+
+        protected class MultipleSelectionFilterTabItem : FilterTabItem<T>
+        {
+            public event Action StateUpdated;
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs:67]: This needs to go. You even already had a bindable `Active` in `TabItem` itself, used it, and then added this anyway...

File: osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            public MultipleSelectionFilterTabItem(T value)
+                : base(value)
+            {
+                Active.BindValueChanged(_ => StateUpdated?.Invoke());
+            }
+
+            protected override bool OnClick(ClickEvent e)
+            {
+                base.OnClick(e);
+                Active.Value = !Active.Value;
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs:78]:
```suggestion
                Active.Toggle();
```

File: osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                RelativeSizeAxes = Axes.X;
+                Height = 15;
+                Spacing = new Vector2(10, 0);
+                AddRange(CreateItems());
+
+                foreach (var item in Children)
+                    item.StateUpdated += updateBindable;
+            }
+
+            protected abstract MultipleSelectionFilterTabItem[] CreateItems();
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs:49]: What is the point? All you care about is the underlying models, not the drawables themselves. How about a

```suggestion
            protected abstract T[] CreateItems();
```

or, even if you decided you want to customise the display text for each model, then

```suggestion
            protected abstract (T model, string description)[] CreateItems();
```

File: osu.Game/Overlays/BeatmapListing/FilterTabItem.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+        protected override void OnDeactivated() => updateState();
+
+        private void updateState()
+        {
+            text.FadeColour(Active.Value ? Color4.White : getStateColour(), 200, Easing.OutQuint);
+            text.Font = text.Font.With(weight: Active.Value ? FontWeight.SemiBold : FontWeight.Regular);
+        }
+
+        private Color4 getStateColour() => IsHovered ? colourProvider.Light1 : colourProvider.Light3;
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/FilterTabItem.cs:74]: does not match web design
```suggestion
        private Color4 getStateColour() => IsHovered ? colourProvider.Light1 : colourProvider.Light2;
```

File: osu.Game/Overlays/BeatmapListing/FilterTabItem.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            updateState();
+        }
+
+        protected override void OnActivated() => updateState();
+
+        protected override void OnDeactivated() => updateState();
+
+        private void updateState()
+        {
+            text.FadeColour(Active.Value ? Color4.White : getStateColour(), 200, Easing.OutQuint);
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/FilterTabItem.cs:70]:
```suggestion
            text.FadeColour(Active.Value ? colourProvider.Content1 : getStateColour(), 200, Easing.OutQuint);
```

also the hover colour actually overrides the active colour on web

**Review by bdach — DISMISSED** [review]

**Comment by bdach** [comment]
I've made several stylistic changes here. @EVAST9919 please look through them whenever you have the chance.

**Review by smoogipoo — APPROVED** [review]
lgtm

**Pull request merged**